src/linear_algebra/projection.lean

```diff
@@ -0,0 +1,179 @@
+  to a submodule `p` along its complement `q`; it is the unique linear map `f : E → p` such that
+  `f x = x` for `x ∈ p` and `f x = 0` for `x ∈ q`.
+* `is_compl_equiv_proj p`: equivalence between submodules `q` such that `is_compl p q` and
+  projections `f : E → p`, `∀ x ∈ p, f x = x`.
+
+We also provide some lemmas justifying correctness of our definitions.
+
+## Tags
+
+projection, complement subspace
+-/
+
+variables {R : Type*} {E : Type*} [ring R] [add_comm_group E] [module R E] (p q : submodule R E)
+
+noncomputable theory
+
+namespace submodule
+
+open linear_map
+
+/-- If `q` is a complement of `p`, then `M/p ≃ q`. -/
+def quotient_equiv_of_is_compl (h : is_compl p q) : p.quotient ≃ₗ[R] q :=
+linear_equiv.symm $ linear_equiv.of_bijective (p.mkq.comp q.subtype)
+  (by simp only [ker_comp, ker_mkq, disjoint_iff_comap_eq_bot.1 h.symm.disjoint])
+  (by simp only [range_comp, range_subtype, map_mkq_eq_top, h.sup_eq_top])
+
+@[simp] lemma quotient_equiv_of_is_compl_symm_apply (h : is_compl p q) (x : q) :
+  (quotient_equiv_of_is_compl p q h).symm x = quotient.mk x := rfl
+
+@[simp] lemma quotient_equiv_of_is_compl_apply_mk_coe (h : is_compl p q) (x : q) :
+  quotient_equiv_of_is_compl p q h (quotient.mk x) = x :=
+(quotient_equiv_of_is_compl p q h).apply_symm_apply x
+
+@[simp] lemma mk_quotient_equiv_of_is_compl_apply (h : is_compl p q) (x : p.quotient) :
+  (quotient.mk (quotient_equiv_of_is_compl p q h x) : p.quotient) = x :=
+(quotient_equiv_of_is_compl p q h).symm_apply_apply x
+
+/-- If `q` is a complement of `p`, then `p × q` is isomorphic to `E`. -/
+def prod_equiv_of_is_compl (h : is_compl p q) : (p × q) ≃ₗ[R] E :=
+begin
+  apply linear_equiv.of_bijective (p.subtype.coprod q.subtype),
+  { simp only [ker_eq_bot', prod.forall, subtype_apply, prod.mk_eq_zero, coprod_apply],
+    -- TODO: if I add `submodule.forall`, it unfolds the outer `∀` but not the inner one.
+    rintros ⟨x, hx⟩ ⟨y, hy⟩,
+    simp only [coe_mk, mk_eq_zero, ← eq_neg_iff_add_eq_zero],
+    rintro rfl,
+    rw [neg_mem_iff] at hx,
+    simp [disjoint_def.1 h.disjoint y hx hy] },
+  { rw [← sup_eq_range, h.sup_eq_top] }
+end
+
+@[simp] lemma coe_prod_equiv_of_is_compl (h : is_compl p q) :
+  (prod_equiv_of_is_compl p q h : (p × q) →ₗ[R] E) = p.subtype.coprod q.subtype := rfl
+
+@[simp] lemma coe_prod_equiv_of_is_compl' (h : is_compl p q) (x : p × q) :
+  prod_equiv_of_is_compl p q h x = x.1 + x.2 := rfl
+
+@[simp] lemma prod_equiv_of_is_compl_symm_apply_left (h : is_compl p q) (x : p) :
+  (prod_equiv_of_is_compl p q h).symm x = (x, 0) :=
+(prod_equiv_of_is_compl p q h).symm_apply_eq.2 $ by simp
+
+@[simp] lemma prod_equiv_of_is_compl_symm_apply_right (h : is_compl p q) (x : q) :
+  (prod_equiv_of_is_compl p q h).symm x = (0, x) :=
+(prod_equiv_of_is_compl p q h).symm_apply_eq.2 $ by simp
+
+@[simp] lemma prod_equiv_of_is_compl_symm_apply_fst_eq_zero {h : is_compl p q} {x : E} :
```

Comment:
I would put `h` explicit here and in the next lemma, since it can't be inferred from the right hand side of the equivalence and you might want to rewrite backwards.

Reply:
Done, waiting for CI.